aep/general/0155/aep.md

```diff
@@ -0,0 +1,106 @@
+# Idempotency
+
+It is sometimes useful for an API to have a unique, customer-provided
+identifier for particular requests. This can be useful for several purposes,
+such as de-duplicating requests from parallel processes, ensuring the safety of
+retries, or auditing.
+
+The purpose of idempotency keys is to provide idempotency guarantees: allowing
+the same request to be issued more than once without subsequent calls having
+any effect. In the event of a network failure, the client can retry the
+request, and the server can detect duplication and ensure that the request is
+only processed once.
+
+## Guidance
+
+APIs **may** add a `string idempotency_key` parameter to request messages
+(including those of standard methods) in order to uniquely identify particular
+requests.
+
+```proto
+message CreateBookRequest {
+  // The parent resource where this book will be created.
+  // Format: publishers/{publisher}
+  string parent = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (google.api.resource_reference) = {
+      child_type: "library.example.com/Book"
+    }];
+
+  // The book to create.
+  Book book = 2 [(google.api.field_behavior) = REQUIRED];
+
+  // A unique identifier for this request. Restricted to 36 ASCII characters.
+  // A random UUID is recommended.
+  // This request is only idempotent if a `idempotency_key` is provided.
+  string idempotency_key = 3 [(google.api.field_info).format = UUID4];
+}
+```
+
+- Providing an idempotency key **must** guarantee idempotency.
+  - If a duplicate request is detected, the server **should** return the
```

Comment:
why not must?

Reply:
Maybe it's not always feasible (or efficient) to do so?

Reply:
actually re-reading and thinking about this a bit - I'm not sure this is even useful guidance.

What benefit is there to the user if this ends up being a copy of the previous successful request? If the response looks similar and the request is idempotent, I don't really see why the content matching matters.

Reply:
Maybe the idea is that if you send the same request twice, it shouldn't matter if you see the first response or the second one.  (Or if you somehow do end up seeing both responses, and they don't match, that could be confusing.)

That is, maybe this is saying NOT to send a second response that is missing important info, because maybe the second response is the only one the consumer receives.